cmd/service-controller/definition_monitor_test.go

```diff
@@ -0,0 +1,565 @@
+				Targets:  []types.ServiceInterfaceTarget{{Name: "dep1", Selector: "label1=value1"}},
+				Origin:   "annotation",
+			},
+			success: true,
+		}},
+		{"http-port-annotation-no-addess-without-selector", newDeployment("dep1", "http", "81", "", 8080, selectorWithoutLabels), result{
+			service: types.ServiceInterface{
+				Address:  "dep1",
+				Protocol: "http",
+				Port:     81,
+				Targets:  []types.ServiceInterfaceTarget{{Name: "dep1", Selector: ""}},
+				Origin:   "annotation",
+			},
+			success: true,
+		}},
+		{"http-port-container-no-address", newDeployment("dep1", "http", "", "", 8080, selectorWithLabels), result{
+			service: types.ServiceInterface{
+				Address:  "dep1",
+				Protocol: "http",
+				Port:     8080,
+				Targets:  []types.ServiceInterfaceTarget{{Name: "dep1", Selector: "label1=value1"}},
+				Origin:   "annotation",
+			},
+			success: true,
+		}},
+		{"http-no-port-no-address", newDeployment("dep1", "http", "", "", 0, selectorWithLabels), result{
+			service: types.ServiceInterface{
+				Address:  "dep1",
+				Protocol: "http",
+				Port:     80,
+				Targets:  []types.ServiceInterfaceTarget{{Name: "dep1", Selector: "label1=value1"}},
+				Origin:   "annotation",
+			},
+			success: true,
+		}},
+		{"http-no-port-with-address", newDeployment("dep1", "http", "", "address1", 0, selectorWithLabels), result{
+			service: types.ServiceInterface{
+				Address:  "address1",
+				Protocol: "http",
+				Port:     80,
+				Targets:  []types.ServiceInterfaceTarget{{Name: "dep1", Selector: "label1=value1"}},
+				Origin:   "annotation",
+			},
+			success: true,
+		}},
+		{"tcp-invalid-port-no-address", newDeployment("dep1", "tcp", "invalid", "", 0, selectorWithLabels), result{
+			service: types.ServiceInterface{},
+			success: false,
+		}},
+	}
+
+	// Iterating through the test table
+	for _, test := range testTable {
+		t.Run(test.name, func(t *testing.T) {
+			service, success := dm.getServiceDefinitionFromAnnotatedDeployment(test.deployment)
+			// Validating returned service
+			assert.Equal(t, test.expected.service.Port, service.Port)
+			assert.Equal(t, test.expected.service.Protocol, service.Protocol)
+			assert.Equal(t, test.expected.service.Address, service.Address)
+			assert.Equal(t, len(test.expected.service.Targets), len(service.Targets))
+			if len(test.expected.service.Targets) > 0 {
+				assert.Equal(t, test.expected.service.Targets[0].Name, service.Targets[0].Name)
+				assert.Equal(t, test.expected.service.Targets[0].Selector, service.Targets[0].Selector)
+			}
+			assert.Equal(t, test.expected.service.Origin, service.Origin)
+			// Validating overall result
+			assert.Equal(t, success, test.expected.success)
+		})
+	}
+
+}
+
+func TestGetServiceDefinitionFromAnnotatedService(t *testing.T) {
+
+	// Types to compose test table
+	type result struct {
+		service types.ServiceInterface
+		success bool
+	}
+
+	type test struct {
+		name     string
+		service  *corev1.Service
+		expected result
+	}
```

Comment:
Another minor comment:
In all the test cases it seems to be a
```type test struct {
		name     string
		service  *corev1.Service
		expected result
	}
```
or a:
```
type test struct {
		name       string
		deployment *v1.Deployment
		expected   result
	}
```

maybe, you can just create two globals:
testDeployment and testService

Reply:
My idea was to keep these inner types contained and private to the
test method, as it is mostly suited to those particular unit tests.

Reply:
I am ok with that!